proposals/2023-08-21-utf8.md

```diff
@@ -0,0 +1,159 @@
+
+* Quoted label name: `{"my.dotted.metric", "http.status"=~"5.."}`
+
+* Classic metrics do not need to be treated differently:
+`sum(rate(regular_metric{region="east"}[5m]))`
+
+* If the user decides to put a classic metric inside the selector, it still needs to be quoted:
+`sum(rate({"regular_metric", region="east"}[5m]))`
+
+* Syntax if the metric has a quote: `sum(rate({'my "quoted" metric', region="east"}[5m]))`
+
+* Recording Rule Syntax: `sum(rate({"my.dotted.metric:errors:rate5m", region="east"}[5m]))`
+
+* Unicode escaping: `{"emoji metric: 💩 or \u1F4A9"}`
+
+* Possible future Native Histogram syntaxes (not being proposed here):
+  `{"my.metric.with.periods", region="east"}.sum`
+  `{"my.metric.with.periods".sum, region="east"}`
+
+* Possible future metric name operator syntax via prefixes (not being proposed here):
+  `{~"a\.regex.*selector.*", region="east"}`
+
+### Implementation
+
+Implementing this feature will require updates to several Prometheus libraries. These updates will be written so that functionality does not change for existing users and builds, and will only be enabled when certain flags are flipped or options turned on.  There will be minimal read-only backwards compatibility for clients that do not support the new supported character set and syntax.
+
+Main tasks:
+
+* Add flags / options to libraries and binaries to enable the new behavior.
+* Change / update code locations that check for a valid metric name, based on the flag setting.
+* Update text generators to generate the new syntax when a metric must be quoted.
+* Update parsers to parse the new syntax.
+* Add an HTTP header so that clients can indicate their support for UTF8.
+* Add a way to escape metric names for clients that do not support the full character set.
+* Update endpoint code to support html-unquoting metric names in http API calls.
+
+### prometheus/common
+
+We will add a boolean argument to model.IsValidMetricName that switches validation logic to allow UTF8 names (and therefore will only check for 0 length). Because this is a public function, all users of this library will need to add an argument to their call sites.
+
+openmetrics_create.go and text_create.go will detect if the metric names require quoting, and if so, use 
+the new syntax when creating the formatted text.
+
+Other repositories that use this check to filter out "bad" metric names will have to be optional via a flag or argument.
+
+### prometheus/client_golang
+
+The golang library checks that metrics have a valid name when NewDesc is called. We can create a new parallel function, NewDescUTF8 (or similar) that will bypass the validity check.
+
+### prometheus/prometheus
+
+All support for UTF8 will be behind a default-off configuration flag. Behavior will not be changed unless this flag is on.
+
+Prometheus itself needs updates to its lexers and parsers to support reading the new exposition syntax and promql syntax:
+
+* openmetricslex
+* promlex
+* generated_parser
+
```

Comment:
Also https://github.com/prometheus/prometheus/blob/main/web/ui/module/codemirror-promql/src/parser/parser.ts (used for syntax highlighting, linting and autocompletion)